DPJTH-1 Add HSM signing sample

datalogics-robl opened the pull request:
#### Changes in this pull request

This adds a functioning sample that signs a PDF using the Luna SA HSM.

This sample does not have unit or integration tests. Those are being added in separate pull requests.
#### Fulfills  [DPJTH-1](https://jira.datalogics.com/browse/DPJTH-1)
#### Checklist for approving this pull request

(PR creator amend this with more conditions if necessary)
- [ ] There are **unit tests** for new/changed code, or a good explanation why this was not possible.
- [x] All **CI builders** have indicated success (Give them a few minutes to notice the pull request.)
- [x] The **Pull request title** has the JIRA issue numbers separated by spaces (if any), a space, and then a short, but descriptive summary.
- [x] **Commit messages** are well formed: [A note about Git commit messages](http://www.tpope.net/node/106)
- [ ] New public packages, classes, and methods are **documented**. (Strongly consider documenting private classes and methods.)
- [ ] Passwords for Datalogics resources or hardware are **not** visible in any code or committed files.
#### Blockers

Add a checkbox for yourself with your **&#64;name** to this list if you are holding this PR open for review. PR Shepherd, please hold off merging this PR until these are all checked:
- [x] _&lt;add your name here with an **@**>_

Commit:
Add resources for the SignDocument sample

Copied from the main Samples repo.

Commit:
Add HsmSignDocument sample

This is a straight copy of the source from SignDocument.java in the main Samples
repo, except for changing the package name.

Commit:
Add static variables needed to talk to the HSM

Commit:
Obtain a Luna SA HsmManager

Commit:
Login to the Luna SA HSM

Commit:
Log success on logging into the HSM, terminate on failure

We can't proceed without a successful login.

Commit:
Obtain and use credentials from the HSM

Call the HSM to get a set of credentials. Note that we need to pass a
JCEProvider to SignatureManager.sign() in order to work, which also requires
obtaining a CryptoContext.

Commit:
Remove dead code for creating credentials locally

We now rely on the HSM manager for this task.

Commit:
Add the PASSWORD field

This is passed to the login manager. At some point soon we may pull this from
a resource file that can be dropped into the current directory, but for now
users need to edit the source file.

Commit:
Update class documentation to describe the sample use

jrowden8 commented on src/main/java/com/datalogics/pdf/hsm/samples/HsmSignDocument.java:
```diff
@@ -0,0 +1,174 @@
+            if (loggedIn) {
+                LOGGER.info("Logged into HSM");
+            } else {
+                LOGGER.info("Failed to log into HSM, exiting");
```
This should probably be a warning message if it fails to login.

datalogics-robl replied:
Good call, I can add that.

jrowden8 commented:
Also I think we

jrowden8 commented:
Whoops accidently closed this PR by mistake.

jrowden8 commented on src/main/java/com/datalogics/pdf/hsm/samples/HsmSignDocument.java:
```diff
@@ -0,0 +1,174 @@
+public final class HsmSignDocument {
+    private static final Logger LOGGER = Logger.getLogger(HsmSignDocument.class.getName());
+
+    private static final String PASSWORD = ""; // <-- Replace this with the HSM partition password
```
We should add functionality so that this sample goes out and fetches the password from some file in the directory path.

Commit:
Revise sample to log a warning if it could not login

Commit:
Catch and rethrow exceptions when logging in

The LunaSA HSM slot manager throws an exception if the password is not
supported. Wrap this in an IllegalArgumentException and rethrow to the caller.

Commit:
Require password for LunaHsmManager to be non-null, non-zero-length.

jrowden8 commented on src/main/java/com/datalogics/pdf/hsm/samples/LunaHsmManager.java:
```diff
@@ -51,12 +53,16 @@ protected LunaHsmManager() {
      */
     @Override
-    public boolean hsmLogin(final String tokenLabel, final String password) {
+    public boolean hsmLogin(final String tokenLabel, final String password) throws IllegalArgumentException {
```
I think we can remove this throws declaration for IllegalArgumentException.

jrowden8 commented on src/main/java/com/datalogics/pdf/security/HsmManager.java:
```diff
@@ -21,8 +21,9 @@
+     * @throws IllegalArgumentException if an argument was invalid
      */
-    boolean hsmLogin(final String tokenLabel, final String password);
+    boolean hsmLogin(final String tokenLabel, final String password) throws IllegalArgumentException;
```
Same thing here.
